pkg/distributor/ingest_limits.go

```diff
@@ -85,57 +85,65 @@ func newIngestLimits(client ingestLimitsFrontendClient, r prometheus.Registerer)
 	return &ingestLimits{
 		client: client,
 		limitsFailures: promauto.With(r).NewCounter(prometheus.CounterOpts{
-			Name: "loki_distributor_ingest_limits_failures_total",
-			Help: "The total number of failures checking ingest limits.",
+			Name: "loki_distributor_ingest_limits_requests_failed_total",
+			Help: "The total number of requests that failed.",
 		}),
 	}
 }
 
-// enforceLimits returns a slice of streams that are within the per-tenant
-// limits, and in the case where one or more streams exceed per-tenant
-// limits, the reasons those streams were not included in the result.
-// An error is returned if per-tenant limits could not be enforced.
-func (l *ingestLimits) enforceLimits(ctx context.Context, tenant string, streams []KeyedStream) ([]KeyedStream, map[uint64][]string, error) {
-	exceedsLimits, reasons, err := l.exceedsLimits(ctx, tenant, streams)
-	if !exceedsLimits || err != nil {
-		return streams, nil, err
+// EnforceLimits checks all streams against the per-tenant limits and returns
+// a slice containing the streams that are accepted (within the per-tenant
+// limits). Any streams that could not have their limits checked are also
+// accepted.
+func (l *ingestLimits) EnforceLimits(ctx context.Context, tenant string, streams []KeyedStream) ([]KeyedStream, error) {
+	results, err := l.ExceedsLimits(ctx, tenant, streams)
+	if err != nil {
+		return streams, err
 	}
 	// We can do this without allocation if needed, but doing so will modify
 	// the original backing array. See "Filtering without allocation" from
 	// https://go.dev/wiki/SliceTricks.
-	withinLimits := make([]KeyedStream, 0, len(streams))
+	accepted := make([]KeyedStream, 0, len(streams))
 	for _, s := range streams {
-		if _, ok := reasons[s.HashKeyNoShard]; !ok {
-			withinLimits = append(withinLimits, s)
+		// Check each stream to see if it failed.
+		// TODO(grobinson): We have an O(N*M) loop here. Need to benchmark if
+		// its faster to do this or if we should create a map instead.
+		var (
+			found  bool
+			reason uint32
+		)
+		for _, res := range results {
+			if res.StreamHash == s.HashKeyNoShard {
+				found = true
+				reason = res.Reason
+				break
+			}
+		}
+		if !found || reason == uint32(limits.ReasonFailed) {
```

Comment:
Since we stopped using `iota`, no valid reason can ever be `0`, meaning we can do a check against the default value of `uint32`.